### PR TITLE
Add TCs for Rancher Integration

### PR DESCRIPTION
**Pull request by albinsun**
### Add managers in `rancher_api`
1. `ClusterDeploymentManager`
1. `ClusterServiceManager`
1. `PVCManager`

### `test_rancher_integration.py`
1. Grouping TCs to classes `TestRKE2` and `TestRKE1`
1. Update `test_create_rke2`: check kube-system deployments `cloud-provider` and `csi-driver`
1. New `test_deploy_nginx`: general deployment
1. New `test_load_balancer_service`: Test service creation and LB routing to nginx deployment
1. New `test_csi_deployment`: deployment with pvc attached

## Verification
Tested on harvester-runtests#394 with
* RKE1-version `v1.24.17-rancher1-1`
* RKE2-version `v1.24.17+rke2r1`

image

### CODE REVIEW

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 1]
```diff
@@ -477,86 +647,67 @@ def test_create_rke1(self, rancher_api_client, unique_name, rancher_wait_timeout
...
             )
-
-    @pytest.mark.rke1
-    @pytest.mark.dependency(depends=["TestRKE::test_create_rke1"])
-    def test_delete_rke1(self, api_client, rancher_api_client, rke1_cluster_name,
-                         rancher_wait_timeout):
-        code, data = rancher_api_client.clusters.get()
+        except AssertionError as e:
+            # teardown
+            rancher_api_client.mgmt_clusters.delete(rke1_cluster['name'])
```
Comment (lanfon72) [inline harvester_e2e_tests/integration/test_rancher_integration.py:687]: we should keep the cluster for debug purpose when the test failed.

**Review by lanfon72 — APPROVED** [review]
In General LGTM

I would suggest not add comments to separate fixtures/test cases, you can use **Code folding** function in your editor.

And for `_polling_for`, **IMO**, it be designed pretty complicated, if the function signature (with types) is not explain what the function going to do, it would need to re-deign.
I know it is very convenient to use the last argument to determine what the function going to do, and javascript use it as a convention to design functions, but It is because function signatures in javascript is not allow keyword only argument, so use last argument is a trade off. (and reflect function's argument is very easy in javascript)

In Python, I would suggest design it like signatures in `multiprocessing`, especially `apply_async` and `map_async`, those signatures elegantly explain how to use it with function and callback and interact while applying multiple functions.

File: harvester_e2e_tests/fixtures/api_client.py [hunk 2]
```diff
@@ -259,31 +264,42 @@ def jumphost_policy(self, allow=True):
 @pytest.fixture(scope="session")
 def polling_for(wait_timeout, sleep_timeout):
     def _polling_for(subject: str,
-                     tester: Callable[..., bool],
-                     testee: Callable, *testee_args, **testee_kwargs):
-        """ Polling expected confition for `wait_timeout`s every `sleep_timeout`s
+                     checker: Callable[..., bool],
+                     poller: Callable, *args,
+                     timeout=wait_timeout):
```
[the remark this answers is not in the document]
Comment (albinsun) [reply harvester_e2e_tests/fixtures/api_client.py:270]: Add a TODO and will try to redesign after v1.2.1 release test.
Thank you.

File: harvester_e2e_tests/integration/test_rancher_integration.py [hunk 3]
```diff
@@ -30,52 +28,10 @@
...
-    warnings.warn(UserWarning(
-        f"Kubernetes version {version} is not in supported list,"
-        f" change to use latest version {latest} instead."
-    ))
-
-    return latest
-
-
-@pytest.fixture(scope='session')
 def vlan_network(request, api_client):
```
[the remark this answers is not in the document]
Comment (albinsun) [reply harvester_e2e_tests/integration/test_rancher_integration.py:32]: Yes, they both define cluster network with `config.getoption('--vlan-nic')`.
But it seems exists before, however will put this for refactoring later.

Do you think it's ok to let `vlan_network`/`vm_network` a session fixture?

**Comment by albinsun** [comment]
tested in `harvester-runtests#396`

**Comment by khushboo-rancher** [comment]
Merging this now, @albinsun Don't forget to create a ticket for the comment https://github.com/harvester/tests/pull/949#discussion_r1348356621 and the TODO task. Thanks

**Pull request merged**